crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,320 @@
+//! Middleware to handle `s3://` URLs to pull artifacts from an S3 bucket
+use std::path::PathBuf;
+
+use async_trait::async_trait;
+use aws_config::BehaviorVersion;
+use aws_sdk_s3::presigning::{PresignedRequest, PresigningConfig};
+use reqwest::{Request, Response};
+use reqwest_middleware::{Middleware, Next, Result as MiddlewareResult};
+use url::Url;
+
+/// S3 middleware to authenticate requests.
+pub struct S3Middleware {
+    config_file: Option<PathBuf>,
+    profile: Option<String>,
```

Comment:
while theoretically you might be able to configure S3 with only env variables like `AWS_CONFIG_PATH` and `AWS_PROFILE`, there might be a scenario where you want to let pixi use a specific profile while still having a different `AWS_PROFILE` (which then would get used by things like terraform in `pixi run`)

that's why we opted to add these two variables (which can be set in `pixi.toml` under `s3.config-file` and `s3.profile`) to override `AWS_CONFIG_PATH` and `AWS_PROFILE`